### PR TITLE
Handle `?`, `*`, `{0,}` and `{0,n}` based repetitions in regexp_replace on the GPU

### LINKED ISSUE #4468: [FEA] Support repetition quantifiers `?` and `*` with regexp_replace
**Is your feature request related to a problem? Please describe.**
We currently fall back to CPU for repetition quantifiers `?` and `*` with `regexp_replace` because the behavior is not consistent with Spark.

**Describe the solution you'd like**
Make the behavior consistent with Spark and enable this on GPU.

**Describe alternatives you've considered**
None

**Additional context**
None

### PR DESCRIPTION
**Pull request by NVnavkumar**
Fixes #4468.

This enables "zero-or-more"-based repetitions in regexp_replace operations that would run the GPU. This was possible due to rapidsai/cudf#10760 being merged and fixing a crash scenario that originally was occurring with these types of repetitions and input that fully matches the regexp.

### CODE REVIEW

**Comment by NVnavkumar** [comment]
build

**Comment by NVnavkumar** [comment]
build

**Comment by andygrove** [comment]
build

File: tests/src/test/scala/com/nvidia/spark/rapids/RegularExpressionTranspilerSuite.scala [hunk 1]
```diff
@@ -288,6 +288,21 @@ class RegularExpressionTranspilerSuite extends FunSuite with Arm {
...
     val inputs = Seq("abc", "a\n\rb", "a\n\u0085b", "a\u2029\u0085b", "a\u2082\rb")
     assertCpuGpuMatchesRegexpFind(pattern, inputs)
+
+  }
+
+  test("replace_replace - ?, *, +, and {0, n} repetitions") {
+    val patterns = Seq("D?", "D*", "D+", "D{0,}", "D{0,1}", "D{0,5}", "[1a-zA-Z]{0,}",
+        "[1a-zA-Z]{0,2}", "A+")
+    val inputs = Seq("SS", "DD", "SDSDSDS", "DDDD", "DDDDDD", "ABCDEFG",
+        "\u000e\u0011\u0006h¹TEST\ud720AA")
```
Comment (andygrove) [inline tests/src/test/scala/com/nvidia/spark/rapids/RegularExpressionTranspilerSuite.scala:298]: This test fails in CI due to the use of the superscript character. I filed https://github.com/NVIDIA/spark-rapids/issues/5521 to investigate this. We do not use characters like this in other tests so I think fixing this is outside of the scope of this PR.

```suggestion
    val inputs = Seq("SS", "DD", "SDSDSDS", "DDDD", "DDDDDD", "ABCDEFG")
```

**Comment by andygrove** [comment]
build

**Review by andygrove — DISMISSED** [review]
LGTM

**Comment by andygrove** [comment]
build

**Comment by andygrove** [comment]
There are new failures since upmerging.

```
- compare CPU and GPU: regexp replace fuzz test with limited chars *** FAILED ***
  javaPattern="$S?a?", cudfPattern="([\n\r\u0085\u2028\u2029]|\r\n)?$S?a?", input='""', cpu="_RE\\PLACE_", gpu="" (RegularExpressionTranspilerSuite.scala:712)
- AST fuzz test - regexp_replace *** FAILED ***
  javaPattern="\u0085?$", cudfPattern="\u0085?([\n\r\u0085\u2028\u2029]|\r\n)?$", input='""', cpu="_RE\\PLACE_", gpu="" (RegularExpressionTranspilerSuite.scala:712)
```

**Comment by andygrove** [comment]
@NVnavkumar could you retarget this for 22.08

**Review by andygrove — APPROVED** [review]
LGTM

**Comment by NVnavkumar** [comment]
build

**Comment by NVnavkumar** [comment]
build

**Comment by sameerz** [comment]
build

**Comment by NVnavkumar** [comment]
build

**Review by anthony-chang — APPROVED** [review]

**Review by andygrove — APPROVED** [review]

**Pull request merged**